haystack/components/rankers/metadata_grouper.py

```diff
@@ -0,0 +1,190 @@
+
+
+    docs = [
+        Document(content="Javascript is a popular programming language", meta={"group": "42", "split_id": 7, "subgroup": "subB"}),
+        Document(content="Python is a popular programming language",meta={"group": "42", "split_id": 4, "subgroup": "subB"}),
+        Document(content="A chromosome is a package of DNA", meta={"group": "314", "split_id": 2, "subgroup": "subC"}),
+        Document(content="An octopus has three hearts", meta={"group": "11", "split_id": 2, "subgroup": "subD"}),
+        Document(content="Java is a popular programming language", meta={"group": "42", "split_id": 3, "subgroup": "subB"})
+    ]
+
+    sample_meta_aggregator = MetaDataGrouper(group_by="group",subgroup_by="subgroup", sort_docs_by="split_id")
+    result = sample_meta_aggregator.run(documents=docs)
+    print(result["documents"])
+    ```
+    """  # noqa: E501
+
+    def __init__(self, group_by: str, subgroup_by: Optional[str] = None, sort_docs_by: Optional[str] = None):
+        """
+        Creates an instance of DeepsetMetadataGrouper.
+
+        :param group_by: The metadata key to aggregate the documents by.
+        :param subgroup_by: The metadata key to aggregate the documents within a group that was created by the
+                            `group_by` key.
+        :param sort_docs_by: Determines which metadata keys are used to sort the documents. If not provided, the
+                             documents within the groups or subgroups are not sorted and are kept in the same order as
+                             they were inserted in the subgroups.
+
+        """
+        self.group_by = group_by
+        self.sort_docs_by = sort_docs_by
+        self.subgroup_by = subgroup_by
+
+    def _buck_into_groups(self, documents: List[Document]) -> Tuple[Dict[str, List], List, List]:
+        """
+        Go through all documents and bucket them based on the 'group_by' value.
+
+        If no 'group_by' value is present in the document, the document is added to a list of documents without a group.
+
+        :param documents:
+        :returns:
+            A tuple with the following elements:
+            - document_groups: A dictionary with the 'group_by' values as keys and the documents as values.
+            - no_group: A list of documents without a 'group_by' value.
+            - ordered_keys: A list of 'group_by' values in the order they were encountered.
+        """
+
+        document_groups: Dict[str, List] = defaultdict(list)
+        ordered_keys: List = []
+        no_group: List = []
+
+        for document in documents:
+            if self.group_by in document.meta:
+                document_groups[str(document.meta[self.group_by])].append(document)
+                if str(document.meta[self.group_by]) not in ordered_keys:
+                    ordered_keys.append(str(document.meta[self.group_by]))
+            else:
+                no_group.append(document)
+
+        return document_groups, no_group, ordered_keys
```

Comment:
the code still seems complex to me.

To simplify, I would propose to return a dictionary instead of a tuple.

We can also introduce an utility method like this:
```python
    def _get_meta_value(self, document: Document, key: str) -> str:
        return str(document.meta.get(key, 'no_group'))
```

The current method could become something like this:
```python
    def _group_documents(self, documents: List[Document]) -> Dict[str, List[Document]]:

        groups: Dict[str, List[Document]] = defaultdict(list)
        
        for doc in documents:
            group_value = self._get_meta_value(doc, self.group_by)
            groups[group_value].append(doc)
            
        return groups
```

`_buck_into_subgroups` could be modified similarly.

Reply:
one of the issues is that you need to accommodate for documents without the specified `group_by` value, if you see the original code those go into a list `no_group` which doesn't exist in your code